backends/p4fmt/attach.cpp

```diff
@@ -0,0 +1,81 @@
+#include "backends/p4fmt/attach.h"
+
+#include "frontends/common/parser_options.h"
+#include "lib/source_file.h"
+
+namespace P4::P4Fmt {
+
+Attach::~Attach() = default;
+
+void Attach::addPrefixComments(NodeId node, const Util::Comment *prefix) {
+    commentsMap[node].prefix.push_back(prefix);
+}
+
+void Attach::addSuffixComments(NodeId node, const Util::Comment *suffix) {
+    commentsMap[node].suffix.push_back(suffix);
+}
+
+bool Attach::isSystemFile(const std::filesystem::path &file) {
+    const std::filesystem::path p4include(p4includePath);
+    return file.parent_path() == p4include;
+}
+
+const Attach::CommentsMap &Attach::getCommentsMap() const { return commentsMap; }
+
+const IR::Node *Attach::attachCommentsToNode(IR::Node *node, TraversalType ttype) {
+    if (node == nullptr || !node->srcInfo.isValid() || clist.empty()) {
+        return node;
+    }
+    std::filesystem::path sourceFile(node->srcInfo.getSourceFile().c_str());
+    if (isSystemFile(sourceFile)) {
+        // Skip attachment for system files
+        return node;
+    }
+
+    const auto nodeStart = node->srcInfo.getStart();
+
+    auto itr = clist.begin();
+    while (itr != clist.end()) {
```

Comment:
Considering performance, are there conditions we could break out from this loop earlier?

That being said, in general I'm okay with leaving performance optimizations as future work.